### PR TITLE
Adding a component for processing, seeded with pre-processing

### PR DESCRIPTION
**Pull request by cmungall**
Transcript of robot-helper chat

User
What are some examples of types of LLM preprocessing, eg distallation
GPT
Large Language Models (LLMs), like the one powering this conversation, often undergo various types of preprocessing to improve their performance, efficiency, or adaptability to specific tasks. These preprocessing techniques can range from methods that simplify the models to techniques that enhance their understanding or generation capabilities. Here are some examples:

Distillation: Knowledge distillation involves training a smaller (student) model to replicate the behavior of a larger (teacher) model. The idea is to compress the knowledge of the large model into a more compact form without significant loss of performance. This can make the model more efficient for deployment in environments with limited computational resources.

Tokenization and Vocabulary Reduction: Before training, text data is broken down into manageable pieces called tokens, which can be words, parts of words, or even characters. Reducing the size of the model's vocabulary (i.e., the set of unique tokens it recognizes) can streamline processing and improve speed, although it might affect the model's ability to understand and generate less common words.

Cleaning and Normalization: This involves removing irrelevant data (e.g., HTML tags from web-scraped content), correcting typos, and standardizing text (e.g., lowercasing, removing extra spaces). These steps help in reducing noise and ensuring that the model learns from clean and consistent data.

Subword Segmentation: Techniques like Byte Pair Encoding (BPE) or SentencePiece break down words into smaller units (subwords or characters), allowing the model to handle a wide range of vocabulary, including rare words, with a fixed-size vocabulary list. This helps in dealing with out-of-vocabulary words and improves the model's ability to generalize.

Data Augmentation: This technique involves artificially expanding the training dataset by making slight modifications to existing data points (e.g., paraphrasing sentences, replacing words with synonyms). Data augmentation can help improve the model's robustness and its ability to generalize from limited data.

Curriculum Learning: In this approach, the model is initially trained on simpler tasks or easier-to-learn data and gradually introduced to more complex tasks. This method can help improve the learning efficiency and final performance of the model.

Transfer Learning: Starting the training process from a model that has already been trained on a related task can significantly reduce training time and improve performance, especially for tasks with limited available data.

Feature Extraction: Before training, specific features or patterns might be extracted from the text to guide the model's learning process. This could include syntactic information, semantic embeddings, or other linguistic features that help the model understand the text better.

These preprocessing techniques can be used in isolation or in combination, depending on the specific requirements of the task at hand and the computational resources available. Each method has its advantages and can significantly impact the performance, efficiency, and applicability of the resulting model.

User
make a robot template for these
GPT
For the task of creating a robot template that covers various types of Large Language Model (LLM) preprocessing techniques, we'll construct a CSV format that includes identifiers for each technique, their names, a brief description, and related concepts or tools that are commonly associated with each technique. Here's how the template would look, following the guidelines for robot templates:

vbnet
Copy code
identifier,name,description,related_concepts
ID,LABEL,A rdfs:comment,A oboInOwl:hasRelatedSynonym

PREPROCESS:1,Distillation,"Knowledge distillation involves training a smaller model to replicate the behavior of a larger model, aiming to compress the knowledge into a more compact form without significant loss of performance.",Knowledge compression|Teacher-student model
PREPROCESS:2,Tokenization and Vocabulary Reduction,"Breaking down text data into manageable pieces called tokens and reducing the model's vocabulary to streamline processing.",Tokenization|Vocabulary size reduction
PREPROCESS:3,Cleaning and Normalization,"Removing irrelevant data, correcting typos, and standardizing text to reduce noise and ensure consistency in the data.",Data cleaning|Text normalization
PREPROCESS:4,Subword Segmentation,"Utilizing techniques like Byte Pair Encoding (BPE) or SentencePiece to break down words into smaller units, allowing the model to handle a wide range of vocabulary with a fixed-size list.",Byte Pair Encoding|SentencePiece
PREPROCESS:5,Data Augmentation,"Expanding the training dataset artificially by modifying existing data points to improve the model's robustness and generalization ability.",Paraphrasing|Synonym replacement
PREPROCESS:6,Curriculum Learning,"Training the model on simpler tasks or easier data first, then gradually introducing more complex tasks to improve learning efficiency and performance.",Sequential learning|Complexity grading
PREPROCESS:7,Transfer Learning,"Starting the training from a model already trained on a related task to reduce training time and improve performance on tasks with limited data.",Pretrained models|Adaptation
PREPROCESS:8,Feature Extraction,"Extracting specific features or patterns from the text before training to guide the model's learning process, including syntactic information or semantic embeddings.",Syntactic information|Semantic embeddings
This template starts with a header row and a template row, as specified in the guidelines for robot templates. The identifier (ID) uniquely identifies each preprocessing technique. The name (LABEL) is the human-readable name of the technique. The description (A rdfs:comment) provides a brief overview of each technique, and related concepts or tools (A oboInOwl:hasRelatedSynonym) list terms closely associated with each technique.

This CSV format allows you to capture key information about various LLM preprocessing techniques in a structured manner that can be further used for documentation, analysis, or as part of a larger ontology on machine learning processes.

User
Add a parent column (SC)
GPT
To include a parent column that categorizes each preprocessing technique under a broader category or concept, I'll add an SC % (subclass) column to the template. This addition helps organize the preprocessing techniques by grouping them under more general categories. For the sake of this example, let's assume these broader categories are:

Data Preparation: For techniques primarily focused on preparing the raw data before training, such as cleaning, normalization, and tokenization.
Model Efficiency: For techniques aimed at making models more efficient, like distillation.
Training Strategies: For techniques that involve specific strategies or methodologies in training, like curriculum learning and transfer learning.
Data Enhancement: For methods that enhance the data or its representation, such as data augmentation and feature extraction.
Here's how the updated template with the parent column included would look:

vbnet
Copy code
identifier,name,description,related_concepts,parent
ID,LABEL,A rdfs:comment,A oboInOwl:hasRelatedSynonym,SC %

PREPROCESS:1,Distillation,"Knowledge distillation involves training a smaller model to replicate the behavior of a larger model, aiming to compress the knowledge into a more compact form without significant loss of performance.",Knowledge compression|Teacher-student model,Model Efficiency
PREPROCESS:2,Tokenization and Vocabulary Reduction,"Breaking down text data into manageable pieces called tokens and reducing the model's vocabulary to streamline processing.",Tokenization|Vocabulary size reduction,Data Preparation
PREPROCESS:3,Cleaning and Normalization,"Removing irrelevant data, correcting typos, and standardizing text to reduce noise and ensure consistency in the data.",Data cleaning|Text normalization,Data Preparation
PREPROCESS:4,Subword Segmentation,"Utilizing techniques like Byte Pair Encoding (BPE) or SentencePiece to break down words into smaller units, allowing the model to handle a wide range of vocabulary with a fixed-size list.",Byte Pair Encoding|SentencePiece,Data Preparation
PREPROCESS:5,Data Augmentation,"Expanding the training dataset artificially by modifying existing data points to improve the model's robustness and generalization ability.",Paraphrasing|Synonym replacement,Data Enhancement
PREPROCESS:6,Curriculum Learning,"Training the model on simpler tasks or easier data first, then gradually introducing more complex tasks to improve learning efficiency and performance.",Sequential learning|Complexity grading,Training Strategies
PREPROCESS:7,Transfer Learning,"Starting the training from a model already trained on a related task to reduce training time and improve performance on tasks with limited data.",Pretrained models|Adaptation,Training Strategies
PREPROCESS:8,Feature Extraction,"Extracting specific features or patterns from the text before training to guide the model's learning process, including syntactic information or semantic embeddings.",Syntactic information|Semantic embeddings,Data Enhancement
With this structure, each preprocessing technique is not only described in terms of what it does and its related concepts but is also categorized under a broader concept that reflects its primary purpose or application area in the context of LLM training and optimization. This hierarchical organization can be valuable for understanding the relationships between different preprocessing techniques and their roles in the broader landscape of LLM development.

User
make the SC column 3rd (after label)
GPT
To rearrange the template so that the parent category (SC %) column appears immediately after the label (name) of each preprocessing technique, here's the updated format:

vbnet
Copy code
identifier,name,parent,description,related_concepts
ID,LABEL,SC %,A rdfs:comment,A oboInOwl:hasRelatedSynonym

PREPROCESS:1,Distillation,Model Efficiency,"Knowledge distillation involves training a smaller model to replicate the behavior of a larger model, aiming to compress the knowledge into a more compact form without significant loss of performance.",Knowledge compression|Teacher-student model
PREPROCESS:2,Tokenization and Vocabulary Reduction,Data Preparation,"Breaking down text data into manageable pieces called tokens and reducing the model's vocabulary to streamline processing.",Tokenization|Vocabulary size reduction
PREPROCESS:3,Cleaning and Normalization,Data Preparation,"Removing irrelevant data, correcting typos, and standardizing text to reduce noise and ensure consistency in the data.",Data cleaning|Text normalization
PREPROCESS:4,Subword Segmentation,Data Preparation,"Utilizing techniques like Byte Pair Encoding (BPE) or SentencePiece to break down words into smaller units, allowing the model to handle a wide range of vocabulary with a fixed-size list.",Byte Pair Encoding|SentencePiece
PREPROCESS:5,Data Augmentation,Data Enhancement,"Expanding the training dataset artificially by modifying existing data points to improve the model's robustness and generalization ability.",Paraphrasing|Synonym replacement
PREPROCESS:6,Curriculum Learning,Training Strategies,"Training the model on simpler tasks or easier data first, then gradually introducing more complex tasks to improve learning efficiency and performance.",Sequential learning|Complexity grading
PREPROCESS:7,Transfer Learning,Training Strategies,"Starting the training from a model already trained on a related task to reduce training time and improve performance on tasks with limited data.",Pretrained models|Adaptation
PREPROCESS:8,Feature Extraction,Data Enhancement,"Extracting specific features or patterns from the text before training to guide the model's learning process, including syntactic information or semantic embeddings.",Syntactic information|Semantic embeddings
This reorganized format emphasizes the categorization of each preprocessing technique right after its name, making it easier to understand the broader context or category to which each technique belongs. This adjustment could be particularly useful for users looking to navigate through the different preprocessing techniques based on their overarching goals, such as data preparation or model efficiency improvements.

User
Make the ID values of the form AIO:CamelCaseLabel
GPT
To adjust the identifier values to the requested format, where each ID is prefixed with "AIO:" followed by a CamelCase version of the label, here's the updated template:

vbnet
Copy code
identifier,name,parent,description,related_concepts
ID,LABEL,SC %,A rdfs:comment,A oboInOwl:hasRelatedSynonym

AIO:Distillation,Distillation,Model Efficiency,"Knowledge distillation involves training a smaller model to replicate the behavior of a larger model, aiming to compress the knowledge into a more compact form without significant loss of performance.",Knowledge compression|Teacher-student model
AIO:TokenizationAndVocabularyReduction,Tokenization and Vocabulary Reduction,Data Preparation,"Breaking down text data into manageable pieces called tokens and reducing the model's vocabulary to streamline processing.",Tokenization|Vocabulary size reduction
AIO:CleaningAndNormalization,Cleaning and Normalization,Data Preparation,"Removing irrelevant data, correcting typos, and standardizing text to reduce noise and ensure consistency in the data.",Data cleaning|Text normalization
AIO:SubwordSegmentation,Subword Segmentation,Data Preparation,"Utilizing techniques like Byte Pair Encoding (BPE) or SentencePiece to break down words into smaller units, allowing the model to handle a wide range of vocabulary with a fixed-size list.",Byte Pair Encoding|SentencePiece
AIO:DataAugmentation,Data Augmentation,Data Enhancement,"Expanding the training dataset artificially by modifying existing data points to improve the model's robustness and generalization ability.",Paraphrasing|Synonym replacement
AIO:CurriculumLearning,Curriculum Learning,Training Strategies,"Training the model on simpler tasks or easier data first, then gradually introducing more complex tasks to improve learning efficiency and performance.",Sequential learning|Complexity grading
AIO:TransferLearning,Transfer Learning,Training Strategies,"Starting the training from a model already trained on a related task to reduce training time and improve performance on tasks with limited data.",Pretrained models|Adaptation
AIO:FeatureExtraction,Feature Extraction,Data Enhancement,"Extracting specific features or patterns from the text before training to guide the model's learning process, including syntactic information or semantic embeddings.",Syntactic information|Semantic embeddings
In this format, each identifier (ID) now uniquely follows the pattern "AIO:" plus the CamelCase version of its corresponding label, ensuring both uniqueness and a clear link to the preprocessing technique it represents. This naming convention makes identifiers more readable and better suited for integration into structured formats or ontologies where clear, concise, and standardized identifiers are necessary.

User
Make the LABEL column all lowercase. Also make sure the the parent class isn't dangling - eg. include a row for "Data Enhancement"
GPT
To modify the template such that the labels are all lowercase and to ensure that each parent class, like "Data Enhancement," has a corresponding row, I'll make the necessary adjustments. This step will include adding entries for each parent category mentioned to avoid any dangling references, thereby ensuring a coherent and complete hierarchy within the dataset.

vbnet
Copy code
identifier,name,parent,description,related_concepts
ID,LABEL,SC %,A rdfs:comment,A oboInOwl:hasRelatedSynonym

AIO:DataPreparation,data preparation,,,"Techniques focused on preparing raw data for training, including cleaning, normalization, and tokenization."
AIO:ModelEfficiency,model efficiency,,,"Techniques aimed at making models more efficient, such as knowledge distillation."
AIO:TrainingStrategies,training strategies,,,"Specific strategies or methodologies employed during model training."
AIO:DataEnhancement,data enhancement,,,"Methods that enhance the training data or its representation, including augmentation and feature extraction."

AIO:Distillation,distillation,model efficiency,"Knowledge distillation involves training a smaller model to replicate the behavior of a larger model, aiming to compress the knowledge into a more compact form without significant loss of performance.",Knowledge compression|Teacher-student model
AIO:TokenizationAndVocabularyReduction,tokenization and vocabulary reduction,data preparation,"Breaking down text data into manageable pieces called tokens and reducing the model's vocabulary to streamline processing.",Tokenization|Vocabulary size reduction
AIO:CleaningAndNormalization,cleaning and normalization,data preparation,"Removing irrelevant data, correcting typos, and standardizing text to reduce noise and ensure consistency in the data.",Data cleaning|Text normalization
AIO:SubwordSegmentation,subword segmentation,data preparation,"Utilizing techniques like Byte Pair Encoding (BPE) or SentencePiece to break down words into smaller units, allowing the model to handle a wide range of vocabulary with a fixed-size list.",Byte Pair Encoding|SentencePiece
AIO:DataAugmentation,data augmentation,data enhancement,"Expanding the training dataset artificially by modifying existing data points to improve the model's robustness and generalization ability.",Paraphrasing|Synonym replacement
AIO:CurriculumLearning,curriculum learning,training strategies,"Training the model on simpler tasks or easier data first, then gradually introducing more complex tasks to improve learning efficiency and performance.",Sequential learning|Complexity grading
AIO:TransferLearning,transfer learning,training strategies,"Starting the training from a model already trained on a related task to reduce training time and improve performance on tasks with limited data.",Pretrained models|Adaptation
AIO:FeatureExtraction,feature extraction,data enhancement,"Extracting specific features or patterns from the text before training to guide the model's learning process, including syntactic information or semantic embeddings.",Syntactic information|Semantic embeddings
By ensuring each parent class has its entry, the template now properly defines and organizes the preprocessing techniques within a structured hierarchy. This inclusion of parent categories as rows, without specific descriptions or related concepts for these category rows, establishes a clear hierarchical structure within the dataset. This approach ensures there are no references to parent classes without corresponding entries, making the dataset coherent and navigable.

### CODE REVIEW

**Comment by caufieldjh** [comment]
It's the fuuuture!
Will need to do some manual edits before merge but this is definitely a good foundation.
A good demonstration for the paper, too.

**Comment by caufieldjh** [comment]
These terms are now in the ROBOT template GSheet but I'm going to merge this in order to build on it further.

**Review by caufieldjh — APPROVED** [review]
